File: graphql-todo-items-java/src/main/java/com/example/subgraphtemplatejavadgsshows/todoitems/TodoItemsDataFetcher.java
package com.example.subgraphtemplatejavadgsshows.todoitems;

import com.example.subgraphtemplatejavadgsshows.DgsConstants;
import com.example.subgraphtemplatejavadgsshows.types.TodoItem;
import com.example.subgraphtemplatejavadgsshows.types.TodoList;
import com.netflix.graphql.dgs.DgsComponent;
import com.netflix.graphql.dgs.DgsData;
import com.netflix.graphql.dgs.DgsDataFetchingEnvironment;
import com.netflix.graphql.dgs.DgsEntityFetcher;

import java.io.IOException;
import java.util.Map;

@DgsComponent
public class TodoItemsDataFetcher {
    private final TodoItemsService itemsService;
    private final TodoItemsMapper mapper;

    public TodoItemsDataFetcher(TodoItemsService itemsService, TodoItemsMapper mapper) {
        this.itemsService = itemsService;
        this.mapper = mapper;
    }

    @DgsEntityFetcher(name = DgsConstants.TODOITEM.TYPE_NAME)
    public TodoItem fetchTodoItem(Map<String, Object> values, DgsDataFetchingEnvironment dataFetchingEnvironment) throws IOException {
        return mapper.todoItemToDgsTodoItem(itemsService.fetchById((String) values.get("id")));
    }

    @DgsEntityFetcher(name = DgsConstants.TODOLIST.TYPE_NAME)
    public TodoList fetchList(Map<String, Object> values) {
        return new TodoList((String) values.get("id"), java.util.List.of(), 0);
    }

    @DgsData(parentType = DgsConstants.TODOLIST.TYPE_NAME)
    public java.util.List<TodoItem> todos(DgsDataFetchingEnvironment dataFetchingEnvironment) throws IOException {
        TodoList list = dataFetchingEnvironment.getSource();
        return itemsService.fetchByListId(list.getId()).stream().map(mapper::todoItemToDgsTodoItem).toList();
    }

    @DgsData(parentType = DgsConstants.TODOLIST.TYPE_NAME)
    public int incompleteCount(DgsDataFetchingEnvironment dataFetchingEnvironment) throws IOException {
        TodoList list = dataFetchingEnvironment.getSource();
        return itemsService.fetchIncompleteCount(list.getId());
    }
}

File: graphql-todo-items-java/src/main/java/com/example/subgraphtemplatejavadgsshows/todoitems/TodoItemsDataFetcher.kt
package com.example.subgraphtemplatejavadgsshows.todoitems

import com.example.subgraphtemplatejavadgsshows.MyContext
import com.example.subgraphtemplatejavadgsshows.types.TodoItem
import com.netflix.graphql.dgs.DgsComponent
import com.netflix.graphql.dgs.DgsData
import com.netflix.graphql.dgs.DgsDataFetchingEnvironment
import com.netflix.graphql.dgs.DgsEntityFetcher
import com.netflix.graphql.dgs.context.DgsContext
import org.openapitools.client.apis.TodoItemControllerApi
import kotlin.collections.List
import kotlin.collections.Map
import kotlin.collections.count
import kotlin.collections.filter
import kotlin.collections.first
import kotlin.collections.listOf
import kotlin.collections.map
import com.example.subgraphtemplatejavadgsshows.types.List as TodoList


@DgsComponent
class TodoItemsDataFetcher {
    
    @DgsEntityFetcher(name = "TodoItem")
    fun fetchTodoItem(values: Map<String, Any>, dataFetchingEnvironment: DgsDataFetchingEnvironment): TodoItem {
        val id = values["id"] as String
        return DgsContext.getCustomContext<MyContext>(dataFetchingEnvironment).fetchTodoItems().first { it.id == id }
    }

    @DgsEntityFetcher(name = "List")
    fun fetchList(values: Map<String, Any>): TodoList {
        return TodoList(values["id"] as String, listOf(), 0)
    }

    @DgsData(parentType = "List", field = "todos")
    fun fetchTodos(dataFetchingEnvironment: DgsDataFetchingEnvironment): List<TodoItem> {
        val list = dataFetchingEnvironment.getSource<TodoList>()
        return DgsContext.getCustomContext<MyContext>(dataFetchingEnvironment).fetchTodoItems().filter { it.listId == list.id }
    }

    @DgsData(parentType = "List", field = "incompleteCount")
    fun fetchIncompleteCount(dataFetchingEnvironment: DgsDataFetchingEnvironment): Int {
        val list = dataFetchingEnvironment.getSource<TodoList>()
        return DgsContext.getCustomContext<MyContext>(dataFetchingEnvironment).fetchTodoItems().count { it.listId == list.id && !it.checked }
    }

}
